### PR TITLE
[DNM] refactor felt252

### PR DESCRIPTION
**Pull request by stevencartavia**
Refactor felt252 #524

### CODE REVIEW

File: src/libfuncs/felt252.rs [hunk 1]
```diff
@@ -218,21 +219,21 @@ pub fn build_binary_operation<'ctx, 'this>(
             // For the initial setup, r0 = PRIME, r1 = a
             // This order is chosen because if we reverse them, then the first iteration will just swap them
             let prev_remainder = start_block
-                .append_operation(arith::constant(context, attr_prime_i512, location))
+                .append_op_result(arith::constant(context, attr_prime_i512, location))
```
Comment (edg-l) [inline src/libfuncs/felt252.rs:222]: When you use append_op_result it removes the need to use `.result(0)?.into()` which is the point of this method. Please fix this on all the other places too, thanks!

**Review by edg-l — CHANGES_REQUESTED** [review]

**Comment by edg-l** [comment]
Please, run git pull locally and fix the formatting issues, with cargo format

File: src/libfuncs/felt252.rs [hunk 2]
```diff
@@ -253,128 +246,94 @@ pub fn build_binary_operation<'ctx, 'this>(
...
-                .result(0)?
-                .into();
+            let rem_times_quo =
+                loop_block.append_op_result(arith::muli(remainder, quotient?, location));
+            let inv_times_quo =
+                loop_block.append_op_result(arith::muli(inverse, quotient?, location));
+            let next_remainder =
+                loop_block.append_op_result(arith::subi(prev_remainder, rem_times_quo?, location));
+            let next_inverse =
+                loop_block.append_op_result(arith::subi(prev_inverse, inv_times_quo?, location));
```
Comment (edg-l) [inline src/libfuncs/felt252.rs:259]: please use the Result unwraps (the `?`) where you call the method, not where you use the result, otherwise we lose the early return and can make things confusing.

TLDR: 

```rust
 let quotient =
                loop_block.append_op_result(arith::divui(prev_remainder, remainder, location));
```

to

```rust
 let quotient =
                loop_block.append_op_result(arith::divui(prev_remainder, remainder, location))?;
```

**Review by edg-l — CHANGES_REQUESTED** [review]

File: src/libfuncs/felt252.rs [hunk 3]
```diff
@@ -253,47 +249,34 @@ pub fn build_binary_operation<'ctx, 'this>(
...
-                    zero,
-                    location,
-                ))
-                .result(0)?
-                .into();
+            let zero = loop_block.append_op_result(arith::constant(
+                context,
+                IntegerAttribute::new(i512, 0).into(),
+                location,
+            ))?;
```
Comment (tcoratger) [inline src/libfuncs/felt252.rs:272]: Here I think you can just use the block ext `const_int` method no?

File: src/libfuncs/felt252.rs [hunk 4]
```diff
@@ -307,43 +290,33 @@ pub fn build_binary_operation<'ctx, 'this>(
...
-                    zero,
-                    location,
-                ))
-                .result(0)?
-                .into();
+            let zero = negative_check_block.append_op_result(arith::constant(
+                context,
+                IntegerAttribute::new(i512, 0).into(),
+                location,
+            ))?;
```
Comment (tcoratger) [inline src/libfuncs/felt252.rs:297]: Here I think you can just use the block ext `const_int` method no?

File: src/libfuncs/felt252.rs [hunk 5]
```diff
@@ -307,43 +290,33 @@ pub fn build_binary_operation<'ctx, 'this>(
...
-                    inverse,
-                    location,
-                ))
-                .result(0)?
-                .into();
+            let prime = negative_check_block.append_op_result(arith::constant(
+                context,
+                attr_prime_i512,
+                location,
+            ))?;
```
Comment (tcoratger) [inline src/libfuncs/felt252.rs:311]: Here I think you can just use the block ext `const_int` method no?

File: src/libfuncs/felt252.rs [hunk 6]
```diff
@@ -217,28 +218,23 @@ pub fn build_binary_operation<'ctx, 'this>(
...
             // This order is chosen because if we reverse them, then the first iteration will just swap them
-            let prev_remainder = start_block
-                .append_operation(arith::constant(context, attr_prime_i512, location))
-                .result(0)?
-                .into();
+            let prev_remainder = start_block.append_op_result(arith::constant(
+                context,
+                attr_prime_i512,
+                location,
+            ))?;
```
Comment (tcoratger) [inline src/libfuncs/felt252.rs:225]: Here I think you can just use the block ext `const_int` method no?

File: src/libfuncs/felt252.rs [hunk 7]
```diff
@@ -217,28 +218,23 @@ pub fn build_binary_operation<'ctx, 'this>(
...
+            let prev_inverse = start_block.append_op_result(arith::constant(
+                context,
+                IntegerAttribute::new(i512, 0).into(),
+                location,
+            ))?;
+            let inverse = start_block.append_op_result(arith::constant(
+                context,
+                IntegerAttribute::new(i512, 1).into(),
+                location,
+            ))?;
```
Comment (tcoratger) [inline src/libfuncs/felt252.rs:237]: Here I think you can just use the block ext `const_int` method no?

File: src/libfuncs/felt252.rs (hunk 3 above)
Comment (stevencartavia) [reply src/libfuncs/felt252.rs:272]: If the block ext `const_int` method is used there it throws an error that 4 arguments are expected and 1 is missing, I am not sure if I can add one more and what would it be.

Comment (tcoratger) [reply src/libfuncs/felt252.rs:272]: Here it should be 

```rust
.const_int(context, location, 0, 512)?
```

I think, 512 being the number of bits for your `i512`, you can check if it works with other examples of the codebase and some refactors being done here for example 

https://github.com/lambdaclass/cairo_native/pull/648

Comment (stevencartavia) [reply src/libfuncs/felt252.rs]: thank you! I updated the PR!

File: src/libfuncs/felt252.rs (hunk 7 above)
Comment (stevencartavia) [reply src/libfuncs/felt252.rs]: Hi, is there an update on this? Can we try to get it merged? 😁

File: tests/tests/felt252.proptest-regressions [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# Seeds for failure cases proptest has generated in the past. It is
```
Comment (tcoratger) [inline tests/tests/felt252.proptest-regressions:1]: I think this file should be removed

File: src/libfuncs/felt252.rs [hunk 9]
```diff
@@ -217,28 +218,11 @@ pub fn build_binary_operation<'ctx, 'this>(
             // Egcd works by calculating a series of remainders, each the remainder of dividing the previous two
             // For the initial setup, r0 = PRIME, r1 = a
             // This order is chosen because if we reverse them, then the first iteration will just swap them
-            let prev_remainder = start_block
-                .append_operation(arith::constant(context, attr_prime_i512, location))
-                .result(0)?
-                .into();
+            let prev_remainder = start_block.const_int(context, location, 0, 512)?;
```
Comment (tcoratger) [inline src/libfuncs/felt252.rs:221]: Here in fact, even better you can use `const_int_from_type`

File: src/libfuncs/felt252.rs [hunk 10]
```diff
@@ -217,28 +218,11 @@ pub fn build_binary_operation<'ctx, 'this>(
...
-            let inverse = start_block
-                .append_operation(arith::constant(
-                    context,
-                    IntegerAttribute::new(i512, 1).into(),
-                    location,
-                ))
-                .result(0)?
-                .into();
+            let prev_inverse = start_block.const_int(context, location, 0, 512)?;
+            let inverse = start_block.const_int(context, location, 1, 512)?;
```
Comment (tcoratger) [inline src/libfuncs/felt252.rs:225]: Here in fact, even better you can use `const_int_from_type`

File: src/libfuncs/felt252.rs [hunk 11]
```diff
@@ -307,43 +274,25 @@ pub fn build_binary_operation<'ctx, 'this>(
...
-                .append_operation(arith::cmpi(
-                    context,
-                    CmpiPredicate::Slt,
-                    inverse,
-                    zero,
-                    location,
-                ))
-                .result(0)?
-                .into();
+            let zero = negative_check_block.const_int(context, location, 0, 512)?;
```
Comment (tcoratger) [inline src/libfuncs/felt252.rs:277]: Here in fact, even better you can use `const_int_from_type`

File: src/libfuncs/felt252.rs [hunk 12]
```diff
@@ -307,43 +274,25 @@ pub fn build_binary_operation<'ctx, 'this>(
...
-            let inverse = negative_check_block
-                .append_operation(arith::select(
-                    is_negative,
-                    wrapped_inverse,
-                    inverse,
-                    location,
-                ))
-                .result(0)?
-                .into();
+            let prime = negative_check_block.const_int(context, location, 0, 512)?;
```
Comment (tcoratger) [inline src/libfuncs/felt252.rs:287]: Here in fact, even better you can use `const_int_from_type`

Comment (stevencartavia) [reply src/libfuncs/felt252.rs:287]: It says in the block_ext file that it should not be used for felt252, should I still use it?

File: .DS_Store
Comment (Oppen) [inline .DS_Store:1]: Delete this file.

**Comment by Oppen** [comment]
@stevencartavia please mark your PRs as drafts until you feel they are ready for review/merge. [DNM] is a nice warning for humans but we can always miss it, the draft marker in GH stops us from being able to merge it by mistake.

File: tests/tests/felt252.proptest-regressions (hunk 8 above)
Comment (Oppen) [reply tests/tests/felt252.proptest-regressions:1]: +1

**Comment by stevencartavia** [comment]
Hi, all the tests passed and the requested file was deleted

File: .DS_Store
Comment (Oppen) [inline .DS_Store:1]: I still need you to delete this file. Adding all `.DS_Store` files to `.gitignore` could be appreciated.

File: src/libfuncs/felt252.rs (hunk 7 above)
Comment (Oppen) [reply src/libfuncs/felt252.rs:237]: If you're aiming for merging already you should remove the "[DNM] " prefix from the title, as it literally means "do not merge", implying you don't consider it ready.

File: .DS_Store
Comment (stevencartavia) [reply .DS_Store:1]: I was going to open a new PR because I struggled updating this one with the main branch and I am seeing that it was already refactored, all the changes from my commits 2 months ago and the comments are already present on the main branch.
